[export] Support preserving calling convention to some modules.

zhxchen17 opened the pull request:
Summary: APS use this feature to swap out some submodules after unflattening.

Test Plan: test_export_preserve_signature

Differential Revision: D48154341



cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

tugsbayasgalan commented on torch/_export/wrappers.py:
```diff
@@ -0,0 +1,116 @@
+            flat_args, in_spec = pytree.tree_flatten((args, kwargs))
+
+            def flat_gm(*flat_args):
+                flat_args = export_tracepoint(*flat_args, kind="module_call_in", path=path)
```
nit: Can you rename this to module_inputs?

tugsbayasgalan commented on torch/_export/wrappers.py:
```diff
@@ -0,0 +1,116 @@
+                args, kwargs = pytree.tree_unflatten(flat_args, in_spec)
+                res = self.inner(*args, **kwargs)
+                flat_res, out_spec = pytree.tree_flatten(res)
+                flat_res = export_tracepoint(*flat_res, kind="module_call_out", path=path)
```
nit: module_outputs

tugsbayasgalan commented on torch/_export/passes/collect_tracepoints_pass.py:
```diff
@@ -0,0 +1,50 @@
+            for node in module.graph.nodes:
+                if node.op != "call_function":
+                    continue
+                if node.target == torch.ops.higher_order.export_tracepoint:
```
This kind of implies that export_tracepoint is public?

zhxchen17 replied:
ok I will add a _

tugsbayasgalan commented on torch/_export/passes/collect_tracepoints_pass.py:
```diff
@@ -0,0 +1,50 @@
+                        else:
+                            assert False
+                        for user in node.users:
+                            assert user.op == "call_function"
```
Hmm why are they getattrs?

zhxchen17 committed:
[export] Support preserving calling convention to some modules. (#106798)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106798

APS use this feature to swap out some submodules after unflattening.

Test Plan: test_export_preserve_signature

Reviewed By: kit1980, tugsbayasgalan

Differential Revision: D48154341

fbshipit-source-id: f56fd1b409c2b9405f98c63e99a86c73c0f0fa6e

zhxchen17 commented:
@pytorchbot merge

zhxchen17 commented:
@pytorchbot merge -f "internal changes has been merged"

zejun-chen commented:
Hi, @zhxchen17 

May i know what does the `APS` mean?
Does it mean the user can use this higher order op to preserve something in `torch.export`?

Thank you.
